### ISSUE TITLE
improve speed of data prep functions

### ISSUE THREAD

**Issue by almartin82** [almartin82/mapvizieR#49]
we have a number of prep functions that need to be applied over each row of our data objects.  a good example is [`fall_spring_me`](https://github.com/almartin82/mapvizieR/blob/a576a85f4b0307d3210a26ffbb6eebb1cae806d6/R/util.R#L300).

`fall_spring_me` [currently](https://github.com/almartin82/mapvizieR/blob/a576a85f4b0307d3210a26ffbb6eebb1cae806d6/R/mapvizieR_object.R#L212) gets applied to our cdf object using a `lapply` call.  in the immortal words of the [r-inferno](http://alyssafrazee.com/vectorization.html) (h/t @alyssafrazee), the apply family "is not vectorization, it is loop-hiding."

other [places](https://github.com/almartin82/mapvizieR/blob/a576a85f4b0307d3210a26ffbb6eebb1cae806d6/R/util.R#L273) in mapvizieR the same thing is getting accomplished via `dplyr::rowwise()` calls.

I'm pretty convinced that this is the wrong approach.  speed really isn't an issue on our test data, which is only ~ 8500 rows long.  but I've been trying out mapvizieR on TEAM's data, and against 68K rows of historic MAP data things get a little balky.  I haven't done a ton of profiling to figure out which calls in the mv object prep are the issue, but regardless, this seems like a good time to attack this problem.

 I re-purposed some johnny miles white code and put together a toy/reproducible set of functions that accomplish the same basic task as [`fall_spring_me()`](https://github.com/almartin82/mapvizieR/blob/a576a85f4b0307d3210a26ffbb6eebb1cae806d6/R/util.R#L300).

put is living as a [gist](https://gist.github.com/almartin82/2fe6a75ba5e19928e350) rather rather than into a vignette because I want to make sure that Travis doesn't end up running this code.

lots of different sources/inspiration for the functions in that gist, but this [post](http://r-de-jeu.blogspot.com/2013/01/search-and-replace-are-you-tired-of.html) on alternatives to nested `ifelse()` was super helpful and provided a lot of grist for the mill.
## Ok so here's the payoff:

```
> n_timings(n=25, test_function="using_join", test_args=list(x=df_test))
25 trials of using_join with mean time of 0.0393 seconds.
min of 0.023 and max of 0.078 seconds.
> 
> n_timings(n=25, test_function="using_if_lapply", test_args=list(x=df_test))
25 trials of using_if_lapply with mean time of 2.3241 seconds.
min of 2.2606 and max of 2.4302 seconds.
> 
> n_timings(n=25, test_function="using_rowwise", test_args=list(x=df_test))
25 trials of using_rowwise with mean time of 2.9456 seconds.
min of 2.6498 and max of 3.5082 seconds.
>
> n_timings(n=25, test_function="using_dplyr_ifelse", test_args=list(x=df_test))
25 trials of using_dplyr_ifelse with mean time of 2.9933 seconds.
min of 2.7422 and max of 3.6562 seconds.
>
> n_timings(n=25, test_function="using_mapvalues", test_args=list(x=df_test$index_col))
25 trials of using_mapvalues with mean time of 0.359 seconds.
min of 0.335 and max of 0.444 seconds.
>
> n_timings(n=25, test_function="using_so_decode", test_args=list(x=df_test$index_col))
25 trials of using_so_decode with mean time of 2.7945 seconds.
min of 2.6322 and max of 3.1022 seconds.
>
> n_timings(n=25, test_function="using_ifelse_chain", test_args=list(x=df_test$index_col))
25 trials of using_ifelse_chain with mean time of 2.731 seconds.
min of 2.5771 and max of 2.9772 seconds.

```

so plyr::mapvalues is decent, using a left_join is the clear winner, and everything else is godawful - 60x worse than the left_join solution.

of course, the join presupposes that you can predict the full scope of your input in advance - the ifelse rules are much more flexible.  but I think this gives me a pretty clear roadmap - use a dplyr_join if at all possible. @chrishaid thoughts?  you can close this 'issue' whenever you think we have the right roadmap.

**Comment by almartin82** [comment]
**185x faster on my machine!**

```
library(mapvizieR)

using_lapply <- function(x) {
  unlist(lapply(x, fall_spring_me)) 
}

using_join <- function(x) {

  #make the df
    #just the grades
    gr_spr <- c(0:12)
    gr_fall <- gr_spr - 0.8
    gr_wint <- gr_spr - 0.5

    with_k <- c('K', gr_spr[2:13])
    #just the labels
    labels_spr <- paste0(with_k, 'S')
    labels_fall <- paste0(with_k, 'F')
    labels_wint <- paste0(with_k, 'W')

  labels_df <- data.frame(
    grade_level_season = c(gr_spr, gr_fall, gr_wint),
    grade_season_label = c(labels_spr, labels_fall, labels_wint),
    stringsAsFactors = FALSE
  )

  input_df <- data.frame(
    grade_level_season = x,
    stringsAsFactors = FALSE
  )

  munge <- left_join(
    x = input_df,
    y = labels_df,
    by = c('grade_level_season' = 'grade_level_season')
  )

  #return as vector
  munge$grade_season_label
}

samp_cdf <- prep_cdf_long(ex_CombinedAssessmentResults)
prepped_roster <- prep_roster(ex_CombinedStudentsBySchool)
samp_cdf$grade <- grade_levelify_cdf(samp_cdf, prepped_roster)
processed_cdf <- process_cdf_long(samp_cdf)

gls <- unique(processed_cdf$grade_level_season)
for_test <- base::sample(gls, 100000, replace = TRUE)  

#are they consistent?
test1 <- using_lapply(x=for_test)
test2 <- using_join(x=for_test)
all.equal(test1, test2)

n_timings(n=20, test_function="using_lapply", test_args=list(x=for_test))
n_timings(n=20, test_function="using_join", test_args=list(x=for_test))

```

**Comment by almartin82** [comment]
closed by pull request #50
